src/test/java/com/google/sps/servlets/UserDataServletTest.java

```diff
@@ -107,17 +104,19 @@ public void testGetMethodWithInvalidId() throws Exception {
     writer.flush();
 
     // Build the correct response output
-    Map<String, Object> correctResponse = new HashMap<>();
-    correctResponse.put(USER_FOUND_PROPERTY, false);
+    ImmutableMap<String, Object> expected = ImmutableMap.of(
+        USER_FOUND_PROPERTY, false);
+    //Map<String, Object> correctResponse = new HashMap<>();
+    //correctResponse.put(USER_FOUND_PROPERTY, false);
 
     // Check that the user wasn't found in datastore
-    verify(mockRequest, atLeast(1)).getParameter(USER_ID_PROPERTY);
-    Assert.assertEquals(gson.toJson(correctResponse), stringWriter.toString().trim());
+    assertThat(gson.toJson(expected)).contains(stringWriter.toString());
```

Comment:
These are strings, so I'm not sure how it works. Is there an "equals" or a "containsExactly" what would work here? Maybe you can convert the output string/json back to a Java object for comparison?

Reply:
[MapSubject documentation](https://truth.dev/api/0.25/com/google/common/truth/MapSubject.html). It seems like I could use the `containsEntry` method for `MapSubjects` in order to compare the mock response to the expected response. This would involve using `Map` (I haven't been able to get `ImmutableMap` to work in this case), and using `gson.fromJson()` to convert the stringwriter's JSON string value. It would look something like this:

`Map<String, Object> actual = (Map) gson.fromJson(stringWriter.toString(), Map.class);
assertThat(expected).containsEntry(USER_FOUND_PROPERTY, actual.get(USER_FOUND_PROPERTY));`

Where there would be an assertion for each property/key. Since isEqualsTo() would be less code, I'm going to stick with that for now, unless you think I should do otherwise.

Reply:
That deserialization looks OK to me. I don't think you need to assert on all properties? If not, individual is better. If you do/want to, then look at something like `containsExactly()`.

Not being immutable here is fine for the test. I do wonder why you have a `(Map)` case if the `fromJson(...)` is specifying to return a map.

Reply:
`MapSubjects` don't seem to provide the `containsExactly()` method. I changed the code to use `containsEntry()` though.